### ISSUE TITLE
GPU delegate: odml.scaled_dot_product_attention composites on Hiera window-attention shapes hard-crash the device (driver hang → watchdog reboot, Pixel 8a)

### ISSUE THREAD

**Issue by john-rocky** [google-ai-edge/LiteRT#8620]
**Repo:** google-ai-edge/LiteRT
**Reproduced on:** Pixel 8a (Mali, ML Drift / LITERT_CL), Android 16; model built with litert-torch 0.10.0, fp16 via ai-edge-quantizer FLOAT_CASTING. **n=1 — the crash reboots the device, so we did not re-run on our shared hardware.**
**Type:** GPU delegate crash — a graph of `odml.scaled_dot_product_attention` composites (window-attention shapes) hard-crashes the device: GPU driver hang → watchdog reboot.

## Summary

Rebuilding the SAM 2.1 Hiera-Tiny **image encoder** with its 12 attentions lowered as
`odml.scaled_dot_product_attention` StableHLO composites produces a model that converts cleanly
(desktop parity corr 1.0, banned ops NONE, ≤4-D) but **hard-crashes the Pixel 8a on its first GPU
run**: the runner process stops responding, adb drops, and the device comes back with `up 0 min`
(watchdog reboot). The decomposed build of the same encoder runs fine on the same device
(fully delegated, numerically correct), as does a composite build of the companion mask decoder —
so the crash is specific to the composite path on this graph.

## Shapes involved

Hiera window attention — the composite inputs are `[B, N, H, d]` with **B up to 1024 windows**
(window tokens N = 64) in the windowed blocks, and `B = 1, N = 4096` in the global-attention blocks
(H = 1–8 heads across stages, d = 96–768 / H). The working decoder composites are `B = 1, N ∈ {8, 4096}`,
H = 8, d = 32 — so high-window-count batches are the prime suspect.

## Repro (⚠ WARNING: expect a device reboot)

1. Take the public conversion script
   https://github.com/john-rocky/LiteRT-Models/blob/main/sam2/scripts/convert_sam2.py
2. Replace the manual attention in `_msa_forward_4d` with the composite form:

```python
from litert_torch.hlfb import StableHLOCompositeBuilder
builder = StableHLOCompositeBuilder(
    name="odml.scaled_dot_product_attention", attr={"scale": float(self.scale)})
query, key, value = builder.mark_inputs(query, key, value)   # [B, N, H, d]
out = torch.nn.functional.scaled_dot_product_attention(
    query.transpose(1, 2), key.transpose(1, 2), value.transpose(1, 2),
    scale=self.scale).transpose(1, 2)
out = builder.mark_outputs(out)
```

3. Convert (encoder only), fp16 FLOAT_CASTING → ~83.6 MB.
4. Run once on a Pixel 8a with `CompiledModel` + GPU accelerator, any 1×3×1024×1024 float input.
   Observed: GPU driver hang, watchdog reboot.

We can hand over the exact converted `.tflite` + input tensors on request.

## Why it matters

The composite path is the *good* lowering (the companion decoder build is correct and faster than the
decomposed one). If it is also the intended future default for plain models (see the companion
litert-torch devx issue), it must not be able to take the whole device down on mainstream
window-attention shapes. A wrong-output or compile-reject failure mode would already be a large
improvement over a reboot.

**Comment by john-rocky** [comment]
Companion converter-side devx issue (why models end up on the decomposed path at all): https://github.com/google-ai-edge/litert-torch/issues/1105

**Comment by shivvaam0001** [comment]
Hi @john-rocky,

Thank you for reporting this issue and providing detailed technical breakdown of the Mali GPU delegate watchdog reboot on the Pixel 8a. We will investigate this further and get back to you.

**Comment by aravindmurali22** [comment]
Hi @jl45G , following up on this thread, are there any recent updates on this? Thanks!
